### ISSUE TITLE
[Feature] [Alpha] Application Priority and Preemption

### ISSUE THREAD

**Issue by RainbowMango** [karmada-io/karmada#5961]
**Summary**
Currently, the `karmada-scheduler` operates based on a First-In-First-Out scheduling order. In some real-world scenarios, like AI training, some jobs are more critical than others and require preferential treatment in terms of scheduling and resource allocation.

We need the feature to enable users to specify the importance of an application. Applications with higher priority are scheduled ahead of other applications with lower priority. When member clusters do not have enough capacity for running a high priority application, the scheduler preempts and removes lower priority applications in order to make room for the high priority application.

**Proposal**:

- [x] [Application Priority and Preemption](https://github.com/karmada-io/karmada/pull/4993)

Author:
- @LeonZh0u
- @seanlaii 
- @wengyao04
- @whitewindmills
- @zclyne (Yifan Zhang)

References:
- https://github.com/karmada-io/karmada/issues/4938
- Proposal in Google Doc: https://docs.google.com/document/d/1MixmgLwnmiRrukyFP25JvE2hqiqB3f_JKT12tXiOsbc/edit#
- https://github.com/karmada-io/karmada/issues/802
- https://github.com/karmada-io/karmada/issues/935

**Iteration Tasks  -- Part-1: Scheduling based on application priority**:
- [x] API change: Introduce `priority class` to `PropagationPolicy` (@seanlaii, https://github.com/karmada-io/karmada/pull/5962)
- [x] API change: Introduce `priority` to `ResourceBinding` (@seanlaii, https://github.com/karmada-io/karmada/pull/5963)
- [x] Feature gate: Introduce a feature gate for this feature. (@RainbowMango, https://github.com/karmada-io/karmada/pull/6151)
- [x] Controller: Build application priority when creating `ResourceBinding` (@LeonZh0u )
- [ ] ~~webhook: Introduce validation to restrict PriorityClass modification~~
- [x] scheduler: Introduce PriorityQueue to replace the current RateLimitedQueue. (@whitewindmills, #6149 )
- [x] E2E tests (@seanlaii, #6306)
- [x] Documentation (@seanlaii, https://github.com/karmada-io/website/pull/822)

**Iteration Tasks  -- Part-2: Preemption in case of member clusters lack of capacity**:
- [ ] API change: Introduce preemptionPolicy to `ResourceBinding`
- [ ] TBD

**Welcome Join us**
If anyone is interested in this feature or would like to participate in its development, please leave a comment in this issue to claim the task you are interested in.

Note that this feature is still under discussion, and the listed iteration tasks might be subject to change.

**Comment by seanlaii** [comment]
Hi @RainbowMango , I can take the first two tasks first. Thank you!

**Comment by seanlaii** [comment]
@assign

**Comment by RainbowMango** [comment]
Sure thanks. Assigned and updated.
By the way, do you know Yifan's GitHub handle? He is the co-author of this proposal, right?

**Comment by whitewindmills** [comment]
@RainbowMango 
in the part-I, we don't need any feature gates.

**Comment by whitewindmills** [comment]
I have already implemented a priority queue demo, so I will continue to work on it. please assign it to me.
/assign
/assign @seanlaii

**Comment by RainbowMango** [comment]
> @RainbowMango in the part-I, we don't need any feature gates.

Not sure about it yet. We probably need a feature gate to isolate the changes. Let's keep it on the list for now.

**Comment by seanlaii** [comment]
> Sure thanks. Assigned and updated.
> 
> By the way, do you know Yifan's GitHub handle? He is the co-author of this proposal, right?

Yes, this is his github: https://github.com/zclyne. Thank you!

**Comment by zhzhuang-zju** [comment]
> webhook: Introduce validation to restrict PriorityClass modification

I'm quite interested in this task. If no one else claims it, I can be assigned it.

**Comment by LeonZh0u** [comment]
I can take on these two tasks:
- Feature gate: Introduce a feature gate for this feature. (if needed)
- Controller: Build application priority when creating ResourceBinding

**Comment by RainbowMango** [comment]
OK Leon. Assigned them to you.
But I think we are still waiting for @kevin-wangzefeng's comments about the feature, so no rush~ I will talk to him.

**Comment by RainbowMango** [comment]
Hi all, @seanlaii @LeonZh0u @whitewindmills Do you guys want to bring this feature into the coming release, which is v1.13.0 scheduled at the end of this month? I know the time is tight, but there is still a chance to have a minimal version of this feature.

**Comment by seanlaii** [comment]
> Hi all, @seanlaii @LeonZh0u @whitewindmills Do you guys want to bring this feature into the coming release, which is v1.13.0 scheduled at the end of this month? I know the time is tight, but there is still a chance to have a minimal version of this feature.

Yes, I really hope we can include this feature in the v1.13.0 release. I understand the timeline is tight, but I’m willing to push for a minimal version within the schedule. Thank you!

**Comment by RainbowMango** [comment]
OK. Let's start with the API parts which are the prerequisites of this feature. I left some comments on your PR, by the way.

**Comment by whitewindmills** [comment]
+1

**Comment by RainbowMango** [comment]
FYI. The API part has been merged, we can proceed with the implementations now. Thank you!

**Comment by seanlaii** [comment]
Hi @zhzhuang-zju , are you planning to work on the Webhook? I can take the E2E tests. Thank you!

**Comment by zhzhuang-zju** [comment]
> Hi @zhzhuang-zju , are you planning to work on the Webhook? I can take the E2E tests. Thank you!

@seanlaii If you're interested in this task, feel free to take it. I can spend more time on the review.

**Comment by RainbowMango** [comment]
Hello @LeonZh0u , Just sent #6151 to add the feature gate because @whitewindmills's PR #6149 relies on it. Hope you don't mind~

**Comment by LeonZh0u** [comment]
I can claim the e2e test ticket as well

**Comment by seanlaii** [comment]
Sounds good to me. I can work on documentation first.

**Comment by RainbowMango** [comment]
/close

As all tasks done. And will open another one for next iteration.

**Comment by vie-serendipity** [comment]
Any update for `PreemptionPolicy`? Or any issue to track?
